### PR TITLE
[RFC] Add vector extension tests

### PR DESCRIPTION
**Pull request by ksco**
Hi,

This draft PR attempts to add tests for the V extension. Since V instructions are much more complex than scalar instructions, handwritten tests don't seem realistic. So I try to use some trivial python scripts to generate tests in batches. Currently about 100 instructions are roughly supported.

## Goal

Since the vector extension specification is frozen for public review, the goal of this PR is to add full 64-bit test support (rv64uv) for the RVV extension. Due to the large number of instructions involved, the workload is relatively large. So I hope to receive some suggestions from community as early as possible.

## About the current patch

The current patch only supports some basic memory and arithmetic instructions, as listed below.

| Instructions                                                 | Type       |
| ------------------------------------------------------------ | ---------- |
| `vl1re8.v, vl1re16.v, vl1re32.v, vl1re64.v`<br />`vl2re8.v, vl2re16.v, vl2re32.v, vl2re64.v`<br />`vl4re8.v, vl4re16.v, vl4re32.v, vl4re64.v`<br />`vl8re8.v, vl8re16.v, vl8re32.v, vl8re64.v` | Memory     |
| `vle8.v, vle16.v, vle32.v, vle64.v`                          | Memory     |
| `vse8.v, vse16.v, vse32.v, vse64.v`                          | Memory     |
| `vadd.vv, vsub.vv, vminu.vv, vmin.vv, vmaxu.vv, vmax.vv, vand.vv, vor.vv`<br />`vxor.vv, vdivu.vv, vdiv.vv, vrem.vv, vremu.vv, vmulhu.vv, vmul.vv`<br />`vmulhsu.vv, vmulh.vv, vsaddu.vv, vrgather.vv` | Arithmatic |
| `vadd.vi, vrsub.vi, vand.vi, vor.vi, vxor.vi, vsaddu.vi, vrgather.vi, vslideup.vi, vslidedown.vx` | Arithmatic |
| `vadd.vx, vsub.vx, vrsub.vx, vminu.vx, vmin.vx, vmaxu.vx, vmax.vx, vand.vx`<br />`vor.vx, vxor.vx, vdivu.vx, vdiv.vx, vrem.vx, vremu.vx, vmulhu.vx, vmul.vx`<br />`vmulhsu.vx, vmulh.vx, vsaddu.vx, vrgather.vx, vslideup.vx, vslidedown.vx` | Arithmatic |
| `vadc.vvm, vsbc.vvm, vmerge.vvm`                             | Arithmatic |
| `vadc.vim, vmerge.vim`                                       | Arithmatic |
| `vadc.vxm, vsbc.vxm, vmerge.vxm`                             | Arithmatic |
| `vfadd.vv, vfsub.vv, vfmin.vv, vfmax.vv, vfsgnj.vv, vfsgnjn.vv, vfsgnjx.vv` | Arithmatic |
| `vfadd.vf, vfsub.vf, vfmin.vf, vfmax.vf, vfsgnj.vf, vfsgnjn.vf, vfsgnjx.vf` | Arithmatic |

Currently, tests cover from 8 to 64 for SEW, and cover from 1 to 8 for LMUL. And the fractional LMUL is not yet supported.

## An example

We use scalar instructions as a benchmark to verify the correctness of vector instructions. For example, the script produces following scalar code for `vadd.vv`:

```assembly
  la a1, tdat     # Load test data address to a1
  la a3, sres     # Load scalar result area address to a3
  la a2, tdat+8   # Load biased test data address to a2
  ld a4, 0(a1)    # Load first element of test data to a4
  ld a5, 0(a2)    # Load first element of biased test data to a5
  add a5, a4, a5  # Add them
  sd a5, 0(a3)    # Store the result to corresponding scalar result area
  ld a4, 8(a1)    # Continue to calculate the remaining elements
  ld a5, 8(a2)
  add a5, a4, a5
  sd a5, 8(a3)
```

Then, use the equivalent vector instruction to write the calculation result to the vector result area.

```assembly
  li t0, -1
  vsetvli t1, t0, e64,m1,ta,ma
  la a2, tdat
  vle64.v v2, (a2)             # Load test data to vs1
  vle64.v v1, (a2)             # Load test data to vd just to fill in some garbage data
  la a2, tdat+8
  vle64.v v3, (a2)             # Load biased test data to vs2

  li t0, 2
  vsetvli t1, t0, e64,m1,ta,ma
  vadd.vv v1, v2, v3           # Add them

  li t0, -1
  vsetvli t1, t0, e64,m1,ta,ma
  la a1, res
  vse64.v v1, (a1)             Store the result to result area
```

Finally, compare whether the two data are the same.

```assembly
  la a2, sres
  TEST_CASE_REG(3, t0, t1, ld t0, 0(a1); ld t1, 0(a2); addi a1, a1, 8; addi a2, a2, 8)
  TEST_CASE_REG(4, t0, t1, ld t0, 0(a1); ld t1, 0(a2); addi a1, a1, 8; addi a2, a2, 8)
```

This method uses scalar instructions to calculate the result data for the comparison at runtime, which can effectively reduce the complexity of the generator.

## Future work

As an experimental PR, I only implemented a small set of instructions. In the future, adding support for more instructions will be a priority. Also, I would increase the coverage of existing tests. For example, support for fractional LMUL, and more complete tail testing. It is also necessary to add some bad-case tests, such as register overlap constraints.

Looking forward to receiving some suggestions!

Thanks!

Yang

### CODE REVIEW

**Comment by quswar-abid** [comment]
Encountered following error while making (`$ make`) for the vadc test:

```
/home/quswarabid/palm/riscv64-unknown-elf-toolchain-10.2.0-2020.12.8-x86_64-linux-ubuntu14/bin/riscv64-unknown-elf-gcc -march=rv64g -mabi=lp64 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -I/home/quswarabid/riscv-tests/isa/../env/p -I/home/quswarabid/riscv-tests/isa/macros/scalar -T/home/quswarabid/riscv-tests/isa/../env/p/link.ld rv64mi/sd-misaligned.S -o rv64mi-p-sd-misaligned
riscv64-unknown-elf-objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data rv64mi-p-sd-misaligned > rv64mi-p-sd-misaligned.dump
/home/quswarabid/palm/riscv64-unknown-elf-toolchain-10.2.0-2020.12.8-x86_64-linux-ubuntu14/bin/riscv64-unknown-elf-gcc -march=rv64gv -mabi=lp64 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -I/home/quswarabid/riscv-tests/isa/../env/p -I/home/quswarabid/riscv-tests/isa/macros/scalar -T/home/quswarabid/riscv-tests/isa/../env/p/link.ld rv64uv/vadc_vim_LMUL1SEW16VL15.S -o rv64uv-p-vadc_vim_LMUL1SEW16VL15
rv64uv/vadc_vim_LMUL1SEW16VL15.S: Assembler messages:
rv64uv/vadc_vim_LMUL1SEW16VL15.S:19: Error: unknown CSR `vcsr'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:94: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:96: Error: unrecognized opcode `vle16.v v2,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:97: Error: unrecognized opcode `vle16.v v1,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:101: Error: unrecognized opcode `vsetvli t1,t0,e8,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:103: Error: unrecognized opcode `vle8.v v0,(a3)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:106: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:107: Error: unrecognized opcode `vadc.vim v1,v2,1,v0'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:110: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:112: Error: unrecognized opcode `vse16.v v1,(a1)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:122: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:124: Error: unrecognized opcode `vle16.v v2,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:125: Error: unrecognized opcode `vle16.v v1,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:129: Error: unrecognized opcode `vsetvli t1,t0,e8,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:131: Error: unrecognized opcode `vle8.v v0,(a3)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:134: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,tu,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:135: Error: unrecognized opcode `vadc.vim v1,v2,1,v0'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:138: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:140: Error: unrecognized opcode `vse16.v v1,(a1)'
make[1]: *** [/home/quswarabid/riscv-tests/isa/Makefile:102: rv64uv-p-vadc_vim_LMUL1SEW16VL15] Error 1
make[1]: Leaving directory '/home/quswarabid/riscv-tests/isa'
make: *** [Makefile:32: isa] Error 2
quswarabid@quswarabid:~/riscv-tests(master)$ 

```

Not sure hwo to fix this! Comments are welcomed.

**Comment by quswar-abid** [comment]
Encountered following error while making (`$ make`) for the vadc test:

```
/home/quswarabid/palm/riscv64-unknown-elf-toolchain-10.2.0-2020.12.8-x86_64-linux-ubuntu14/bin/riscv64-unknown-elf-gcc -march=rv64g -mabi=lp64 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -I/home/quswarabid/riscv-tests/isa/../env/p -I/home/quswarabid/riscv-tests/isa/macros/scalar -T/home/quswarabid/riscv-tests/isa/../env/p/link.ld rv64mi/sd-misaligned.S -o rv64mi-p-sd-misaligned
riscv64-unknown-elf-objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data rv64mi-p-sd-misaligned > rv64mi-p-sd-misaligned.dump
/home/quswarabid/palm/riscv64-unknown-elf-toolchain-10.2.0-2020.12.8-x86_64-linux-ubuntu14/bin/riscv64-unknown-elf-gcc -march=rv64gv -mabi=lp64 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -I/home/quswarabid/riscv-tests/isa/../env/p -I/home/quswarabid/riscv-tests/isa/macros/scalar -T/home/quswarabid/riscv-tests/isa/../env/p/link.ld rv64uv/vadc_vim_LMUL1SEW16VL15.S -o rv64uv-p-vadc_vim_LMUL1SEW16VL15
rv64uv/vadc_vim_LMUL1SEW16VL15.S: Assembler messages:
rv64uv/vadc_vim_LMUL1SEW16VL15.S:19: Error: unknown CSR `vcsr'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:94: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:96: Error: unrecognized opcode `vle16.v v2,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:97: Error: unrecognized opcode `vle16.v v1,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:101: Error: unrecognized opcode `vsetvli t1,t0,e8,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:103: Error: unrecognized opcode `vle8.v v0,(a3)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:106: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:107: Error: unrecognized opcode `vadc.vim v1,v2,1,v0'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:110: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:112: Error: unrecognized opcode `vse16.v v1,(a1)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:122: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:124: Error: unrecognized opcode `vle16.v v2,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:125: Error: unrecognized opcode `vle16.v v1,(a2)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:129: Error: unrecognized opcode `vsetvli t1,t0,e8,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:131: Error: unrecognized opcode `vle8.v v0,(a3)'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:134: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,tu,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:135: Error: unrecognized opcode `vadc.vim v1,v2,1,v0'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:138: Error: unrecognized opcode `vsetvli t1,t0,e16,m1,ta,ma'
rv64uv/vadc_vim_LMUL1SEW16VL15.S:140: Error: unrecognized opcode `vse16.v v1,(a1)'
make[1]: *** [/home/quswarabid/riscv-tests/isa/Makefile:102: rv64uv-p-vadc_vim_LMUL1SEW16VL15] Error 1
make[1]: Leaving directory '/home/quswarabid/riscv-tests/isa'
make: *** [Makefile:32: isa] Error 2
quswarabid@quswarabid:~/riscv-tests(master)$ 
```

Not sure hwo to fix this! Comments are welcomed.

**Comment by quswar-abid** [comment]
1. Changing `VLEN` changes the configuration of `VL` in generated tests. The selection of what `VL` is generated seems arbitrary. I found that immediates generated for some instructions exceed the `12-bit` limit on load and store instructions. It happened to me at `VLEN=4K` and so invalid tests are generated.

2. The tests are generated in all possible ways `SEW` and `LMUL` can appear, the `VL` is selected on boundaries of a specific number of bits. What is the reason behind this? I'd really appreciate your answer on this one too.

File: codegen/vector/main.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+
+    def __str__(self):
+        nbytes = self.option.nbytes()
+        nquads = nbytes // 8
+        remaining_bytes = nbytes % 8
+        test_data_bytes = (VLEN * self.option.lmul) // 8 + 8
+        test_data = generate_test_data(test_data_bytes)
+        test_data_str = "\n".join([f"  .quad 0x{e:x}" for e in test_data])
+
+        start = inc(2)
```
Comment (quswar-abid) [inline codegen/vector/main.py:122]: Tests cases are registered from case#3, Isn't it supposed to be like `start = inc(2)` to generate proper number of cases?

**Comment by ksco** [comment]
We actually don't use those unit tests anymore, we have a new way (and a better way IMHO) to generate tests. You can find these tests at link [1]. The code used to generate these tests is in link [2].
I'll submit a new PR when it's ready, so I think I just close this one.

[1] https://github.com/ksco/riscv-tests/tree/spike
[2] https://github.com/ksco/riscv-isa-sim/tree/rvv-tests

**Comment by jerryz123** [comment]
@ksco is there any update on those tests? I'd be happy to assist in upstreaming.

**Pull request closed without merging**